### ISSUE TITLE
Idea: restructure WP Rig to support a source/build style

### ISSUE THREAD

**Issue by ataylorme** [wprig/wprig#129]
## Issue Overview
The current approach of editing files in a nested `dev` directory within the production version of the theme surfaces some issues.

* Deleted files are not removed from the theme (#101)
* Files not included in production are included (#118)
* It is not intuitive to create new WordPress PHP files in the `dev` directory

## Expected Behavior
When working with other software that has to be built for production there is typically a source and a build directory without the source directory and files being included in/nested in the build directory.

## Possible Solution
Rather than having a `dev` directory nested in the WP Rig theme I propose WP Rig, as hosted on GitHub, be a development theme that is never actually activated. The `dev` directory is removed and all source files, task processes/automation, etc. are in the top level WP Rig development theme.

Gulp is then used to generate a production version of the theme. This would exclude non-proudction assets, automation scripts/tooling, tests, etc. The production version of the theme is what would be activated.

When developing all work is done on the development version of the theme which will update the activated production theme. The existing `bundle` command can also be used to generate a `.zip` of the production ready theme.

This keeps the source/development files and production files separated and well organized, more closely matching a traditional source/build model.

**Comment by mor10** [comment]
Some background on why the folder structure is the way it is right now:

From a user standpoint, two of the main goals for WP Rig were:
a) for it to work within a development environment out of the box meaning once you added WPRig to the `/wp-content/themes/` folder and ran the process, a theme would emerge.
b) the theme development folder (currently `/dev/`) would be the only place most developers made changes to code, and that this folder look as close to a "normal" theme as possible.

These two principles are core to the entire project and should not be strayed from. As long as your proposal does not collide with them, I am interested in seeing where this idea leads.

To be quite honest, I'm not entirely sure what you're proposing here, so an example of some sort would be useful ;o)

**Comment by ataylorme** [comment]
I'm proposing not having a nested `dev` directory in the active theme. Instead have `wp-content/themes/wp-rig` be the place to develop and `wp-content/themes/my-theme` be the artifact that WP Rig produces. 

Items not necessary in production, such as `composer.json`, would be excluded from the production theme. Conversely production only-items, like minified CSS, would be excluded from the development theme.

A lot of what is in `dev` now, the files developers work on, would be in the development theme.

I think another benefit to this approach is a developer could create a theme with WP Rig and just hand off the production theme to a client and the client would not need WP Rig knowledge to use the built theme.

**Comment by ataylorme** [comment]
> for it to work within a development environment out of the box meaning once you added WPRig to the /wp-content/themes/ folder and ran the process, a theme would emerge.

This would still be true. Out of the box after running the build you would just get a usable theme in a different directory than WP Rig

> the theme development folder (currently /dev/) would be the only place most developers made changes to code, and that this folder look as close to a "normal" theme as possible.

I would argue that a source theme would look closer to a _normal_ theme as the templates, `fuctions.php`, etc would not be nested in a `dev` directory until the first build is ran

**Comment by mor10** [comment]
that makes sense... ish. I think my brain is too elsewhere at the moment to think it through completely. Give me a few days

**Comment by felixarntz** [comment]
One important thing that we should consider as well when thinking about this is that **"building" PHP during development is far from optimal and completely redundant**. This is a problem with the current approach, and having completely separate theme directories wouldn't solve that either.

"Building" PHP means simply copying PHP files which has no benefit whatsoever. On the contrary, now when you make a change, you need to run a script (or have it run via watch task) to see your changes reflected. This makes sense for CSS and JS, but has no point in PHP. For WP Rig specifically, to be fair there is the one benefit regarding the string replacements made in PHP files. However, I think that could be solved in another way, especially as you usually decide on a name once and after that it should never be required to make those string replacements again.

WordPress core currently has a build step that also involves PHP, however that was created with CSS and JavaScript in mind, without too many thoughts lost about PHP. Because of this, it is a pain contributing to PHP in WordPress core at the moment, and it has already been decided to find a way to partly revert this, to get the best of both worlds. I'm not entirely sure what this is going to look like, but it might very likely be something to follow (or at least be inspired by) for solving the quite similar problem with WP Rig.

Something I could see as a solution to all of the mentioned problems:
* Make WP Rig a development theme, similar as @ataylorme has suggested, with a file structure that is common for WordPress themes, where the root contains the source, not a `dev` directory.
* Different from his proposal, build the CSS and JS assets into the same WP Rig development theme. The built files should be excluded from git, but still be in the same theme, so that you can simply run the development theme. For example, built files could be in `assets/css` and `assets/js` while their source files are in `assets/src/css` and `assets/src/js`.
* When you actually want to build the theme for production, there is another build step that simply runs all the "development build" steps, and then copies everything over to a separate theme directory (also similar to what @ataylorme suggested). At this point, string replacements should be applied (i.e. the development theme should always continue to use `WP Rig` naming, without any string replacements being part of the "development build" process).
* As a summary, as far as I understood him, this is close to what @ataylorme suggested, however it should be possible to actually run the development theme as well. The production theme should only be that, for production, while you can develop and test in the `wp-rig` theme entirely.
* Benefits of this approach:
    * You can use the same theme you develop in to test your changes during development.
    * People create PHP files in the theme root as usual.
    * PHP files do not require a build step in order to test the theme during development.
    * CSS and JS files continue to be built, no changes here really.
    * Because the replacements are only applied to the separate production-theme directory, we don't need to worry about it during development, and the development theme should always continue to use `WP Rig` naming.
    * As such, developers can easily have a custom fork of `WP Rig` for development, and they can always simply merge in the latest upstream changes. This is another major benefit over Underscores, which you at the moment just checkout once, make replacements and then never merge in its changes again, unless you like going through a manual very painful one-by-one procedure.

**Comment by ataylorme** [comment]
> One important thing that we should consider as well when thinking about this is that "building" PHP during development is far from optimal and completely redundant.

Thank you for the insightful response @felixarntz. I agree with your amendment, which I basically see as changing the proposal to:
1. Moving PHP files out of the `dev` subdirectory
1. Using dedicated `src` and `build` directories for assets that need to be built for production (css, js, images, etc.).

Our current production bundle process could even be modified to use a temporary directory. All production files (PHP, css, js, images, translation files, etc.) would be copied there and then string replacement would happen for the PHP files. This would leave the PHP files in the development version of the theme unmodified. We could also exclude source files and documentation, such as the readme and contributor files, from the production build. Since the development PHP files do not get strings replaced, as you stated, WP Rig updates can more easily be merged.

I also propose that during normal development, as gulp is watching for changes to source files, we build unminified assets. This will allow for easier debugging. Minified assets can be built during a production run.

**Comment by mor10** [comment]
If at all possible, I'd like to keep the dev theme as "intact" as possible. Once we start abstracting things by placing them in different folders from where they end up, it becomes much harder for old-school WP developers to follow what's going on, and we step too far away from 99% of the tutorials out there. We have to introduce some sort of ramp, not a vertical cliff.

**Comment by ataylorme** [comment]
> If at all possible, I'd like to keep the dev theme as "intact" as possible. Once we start abstracting things by placing them in different folders from where they end up, it becomes much harder for old-school WP developers to follow what's going on, and we step too far away from 99% of the tutorials out there

I agree with this sentiment but would argue the current setup of editing files in a `dev` subdirectory already violates it. The WordPress standard is to edit PHP files in the location, not elsewhere, which this proposal rectifies.

As for other assets (CCS, JS, images, etc.), a source and production distinction with a build step is commonplace, both in WordPress theme development and the larger development ecosystem.

I am not clear on how this proposal steepens the learning curve.

**Comment by mor10** [comment]
I need to see an example. Can you sketch out a rough file/folder tree or something so I get a general idea of where you see this ending up.

**Comment by ataylorme** [comment]
The root directory would look like this:

```
404.php
BrowserSync
CONTRIBUTING.md
LICENSE
README.md
comments.php
composer.json
composer.lock
css
dev
footer.php
functions.php
gulp
gulpfile.babel.js
header.php
images
inc
index.php
js
languages
node_modules
package-lock.json
package.json
phpcs.xml.dist
phpunit.integration.xml.dist
phpunit.xml.dist
pluggable
screenshot.png
sidebar.php
style.css
template-parts
vendor
verbose
```

The dev directory would look like this:
```
dev
├── config
│   ├── cssVariables.json
│   └── themeConfig.js
├── css
│   ├── comments.css
│   ├── content.css
│   ├── editor-styles.css
│   ├── front-page.css
│   ├── sidebar.css
│   └── widgets.css
├── images
│   └── placeholder.svg
├── js
│   ├── customizer.js
│   ├── navigation.js
│   └── skip-link-focus-fix.js
├── optional
│   ├── archive.php
│   ├── attachment.php
│   ├── category.php
│   ├── custom-page-template.php
│   ├── custom-post-template.php
│   ├── front-page.php
│   ├── page.php
│   ├── search.php
│   ├── single.php
│   └── singular.php
├── style.css
└── tests
    └── phpunit
        ├── bootstrap-common.php
        ├── bootstrap-integration.php
        ├── bootstrap-unit.php
        ├── framework
        │   ├── Integration_Test_Case.php
        │   └── Unit_Test_Case.php
        ├── integration
        │   ├── Hooks_Tests.php
        │   └── Theme_Support_Tests.php
        └── unit
            └── Setup_Tests.php
```

By default, folks can edit `functions,php`, `header.php`, `footer.php`, etc. in the root theme directory.  PHP files in `dev/optional` can be moved to the root directory if they are needed but largely serve as examples.

The asset (CSS, JS, images, etc.) source files get edited in the `dev` directory and then get compiled to their respective folders in the root directory.

gulp can also watch for those changes. We could also watch for changes to PHP files and do the string replacement on file save as well (personally I could go either way on that).

Then to create a production bundle the gulp task would rebuild all the assets in their minified version and create a `.zip` that excludes files in the `dev` directory as well as the README and other non-essential items.

**Comment by mor10** [comment]
OK. So you work on PHP files in the root directory and CSS and JS files in the `/dev/` directory? My concern with this is the CSS and JS files live in two places: both root and under `/dev/`, and I have a suspicion many developers will work on the files in root instead of the `/dev/` directory which would not be functional. The reason for putting everything in the `/dev/` directory was to avoid people having to work on files in two different locations. How do we resolve that?

**Comment by ataylorme** [comment]
> I have a suspicion many developers will work on the files in root instead of the /dev/ directory which would not be functional.

I think this will be a concern as long as we have source and production files for CSS/JS.

**Comment by luminuu** [comment]
I think putting the CSS/JS files under a separate folder like `assets`, as Felix suggested, could make it easier to understand that these are the built files.

I like the idea of restructuring WP rig the way Andrew and Felix have suggested it. Then it would be quite close to the theme structure I have developed in the past years. I personally think this is a good approach to integrate modern development workflows into the file structure everyone knows from Underscores or the Twenty themes.

**Comment by felixarntz** [comment]
I agree with @ataylorme and @luminuu that having PHP files editable in the root and buildable assets through some subdirectory is as close as it could get to how themes are typically written for WordPress.

The unnecessary build step for PHP makes the development process less intuitive for no good reason. CSS and JS _are_ buildable assets though, so it's typically expected for those to be edited in one place and then built into the other (as long as these two locations are immediately obvious).

Developers who come from https://github.com/Automattic/_s would already be familiar with this workflow. Sure, we want to do some things in a more modern way, but in this case I'd argue the way themes have been doing this for a long time is already solid enough.

An `assets/css` and `assets/js` (or possibly even `assets/dist/css` and `assets/dist/js`), in combination with `assets/src/css` and `assets/src/js` make the build direction of those files obvious at first glance.

With the extra build step of _actually_ getting the theme ready for production, by copying files to an additional theme directory and performing the replacements, we would still provide a major benefit over Underscores, where the replacements are made _in_ the actual code you edit, making it impossible to automatically merge in upstream changes of the starter theme later.

In regards to keeping the `dev` directory because of the tutorials, I don't think that is a good reason for the long-term success of this project, and by releasing a 2.0 (see https://github.com/wprig/wprig/issues/121#issuecomment-424252011) with all the recent changes (most of which were of structural nature as well) and this one here we would make clear that those who follow the tutorials should continue to use the 1.0 branch. We should write a document that is available before the release where we explain the most significant changes and how people can migrate from 1.0 to 2.0.

**Comment by ataylorme** [comment]
> We should write a document that is available before the release where we explain the most significant changes and how people can migrate from 1.0 to 2.0.

Seems like another good use case for the wiki though I'm not convinced folks already using `1.0` should migrate. Explaining the differences so that folks can use `2.0` in their next project is a good idea but I don't see the benefit in migrating an existing project from `1.0` to `2.0`. It would be a lot of manual effort to move files to the new directory structure, replace the gulp process, etc.

**Comment by ataylorme** [comment]
> An assets/css and assets/js (or possibly even assets/dist/css and assets/dist/js), in combination with assets/src/css and assets/src/js make the build direction of those files obvious at first glance.

Agreed. I think source and build directories are a common practice and one most developers are familiar with (sass has been actively maintained for over 10 years).

**Comment by ataylorme** [comment]
> An assets/css and assets/js (or possibly even assets/dist/css and assets/dist/js), in combination with assets/src/css and assets/src/js make the build direction of those files obvious at first glance.

The main things I am advocating for are
1. Have the PHP files worked on in the root directory rather than the `dev` directory
    - String replacement does not occur to these files directly
1. Have a clear production bundle step (PHP string replacements, minified assets, excluding unnecessary files, etc.)
1. We separate source and built assets

I am indifferent about where the source and built assets live, whether we build the production version of the theme to another `wp-content/themes` directory, a temporary directory or solely in a `.zip` file, and what remains in the `dev` directory.

If we agree on principles then we can move forward with making decisions around the other items. I am okay with that being a decision @mor10 and @bamadesigner make rather than having a large group debate.

**Comment by mor10** [comment]
Move forward with the idea as proposed as `v2.0`. You have my blessing. The `v1.x` branch will stay intact and be the focus of the tutorials. As we built out `v2.0` we'll have to provide some documentation on how it works, and this documentation should target new users. I agree with @ataylorme that having people upgrade from `v1` to `v2` mid-production won't make any sense anyway.

**Comment by ataylorme** [comment]
> Move forward with the idea as proposed as v2.0

@mor10 do you want to create a new `v2.0` branch or keep things in the existing `v1.1` branch? I think similarly to #121 we should decide what will be in `v2.0` aside from these structural changes.

**Comment by mor10** [comment]
Let's make a new `v2.0` branch based on the current `v1.1` branch and then put `v1.1` to pasture. That way there is no confusion.

I've created two new milestones: `v2.0` and `v2.1` and started sorting what goes where. All the CSS refactoring stuff has been punted to `v2.1` as has the SVG issue.

All the `v2.0` issues have been added to a Project so we can see where everything is. Most of them are fairly minor and I suspect several have already been dealt with or can be fixed by adding one or two lines of code somewhere. If necessary / desired, we can split this into several different projects to create a clear separation of concerns.

**Comment by mor10** [comment]
`v2.0` branch is now available. It's a duplicate of `v1.1` branch as of a few minutes ago. Have at it!

**Comment by mor10** [comment]
#121 is closed. Let's capture all discussion of what features do or do not go into `v2.0` in #131.

**Comment by felixarntz** [comment]
Agreed, it’s unlikely people migrate an already active theme from v1 to v2. I think docs about the changes and migration hints would still be useful as there are probably also many folks who are only experimenting with WP Rig so far, and those people might want to move to the new version.

**Comment by mor10** [comment]
I just realized something which may or may not be a problem: One of my original reasons for keeping the PHP files in the `/dev/` folder and running the Gulp process on them was to allow the developer to use the `wprig` prefix on functions and have them automatically replaced on build. How do we preserve this behavior in the new paradigm @ataylorme?

**Comment by ataylorme** [comment]
> Correct me if I'm wrong @ataylorme!

Nope, sounds right to me. The only thing different in the production build vs the development build will be production assets (minified css/js, optimized images, translation files, etc) and the string replacements in assets and PHP files.

As @felixarntz said the development build will very much be usable. In an ideal scenario, the development build will be used locally and a CI service can handle the production build, deployment, and automated testing.

**Comment by mor10** [comment]
Makes sense. Just making sure all our i's are dotted.

**Comment by mor10** [comment]
Decisions:

- `/dev/` directory is deprecated. Files from the directory will be moved down one level.
- Working directory for JavaScript files will be `/js/src/` which compiles to `/js/`.
- Working directory for CSS files will be `/css/src/` which compiles to `/css/`.
- Images stay in `/images/` and are only processed on production / export.
- Optional PHP files stay in `/optional/`.

**Comment by ataylorme** [comment]
> `/dev/` directory is deprecated. Files from the directory will be moved down one level.

Do you mean _up_ one level to the root?

**Comment by mor10** [comment]
Development process (`npm run dev`):
- Compiles unminified CSS from `/css/src/` to `/css/` passing through PostCSS and StyleLint
- Compiles unminified JS from `/js/src/` to `/js/` passing through Babel and ESLint
- Checks PHP files against PHPCS
- Boots up / refreshes BrowserSync
- Watches for changes to all theme files

Production process (currently `npm run bundle`):
- Compiles minified CSS from `/css/src/` to `/css/` passing through PostCSS and StyleLint
- Compiles minified JS from `/js/src/` to `/js/` passing through Babel and ESLint
- Checks PHP files against PHPCS
- Runs prefix renaming process on all files (PHP, JS, other)
- Runs image optimization process on `/images/` folder
- Runs translate process on everything
- Saves output files to separate folder / zip archive depending on setting in `/config/themeConfig.json`
- Ignores build process files and `/optional/` folder during above save process

**Comment by ataylorme** [comment]
> Ignores build process files and /optional/ folder during above save process

Should we go with a whitelist for what is included? Something like all PHP files (excluding those in `optional`), `css` and `js` (excluding their `src` subdirectories), `images`

I think that would be easier to understand and maintain than a blacklist to exclude things like the WP Rig README, etc.

**Comment by mor10** [comment]
Absolutely. Similar to the inverted `.gitignore`. Much easier to manage.

**Comment by mor10** [comment]
One thing we have to figure out is how to manage custom page/post template files in the `/optional/` folder. They will show up as live templates even when they are in that folder when the developer uses the `/wprig/` directory as a live theme.

**Comment by ataylorme** [comment]
> One thing we have to figure out is how to manage custom page/post template files in the `/optional/` folder. They will show up as live templates even when they are in that folder when the developer uses the `/wprig/` directory as a live theme.

We can use the [`theme_scandir_exclusions` filter](https://core.trac.wordpress.org/browser/tags/4.9/src/wp-includes/class-wp-theme.php#L1155)

**Comment by ataylorme** [comment]
> Can we do that in development and remove it for production?

Possibly. We could put a large PHP comment in `functions.php` like _Development items only below this line_ and then have the production build trim everything below that comment.

I am not convinced it is necessary though as we will already be removing the `optional` directory from the production bundle so the addition to the filter is harmless.

**Comment by mor10** [comment]
let's make it as simple as possible first (use the filter and leave it in place) and then see if we need to do something weird later on

**Comment by ataylorme** [comment]
I would like to take all the decision here and make a new issue with the work to be done so there is a clear checklist on when work is complete.

**File Restructuring**
- [ ] Move all files and directories in `dev` up one level to the root directory
    - [ ] Exception: `dev/css` should become `/css/src`
    - [ ] Exception: `dev/js` should become `/js/src`
- [ ] use the [`theme_scandir_exclusions` filter](https://core.trac.wordpress.org/browser/tags/4.9/src/wp-includes/class-wp-theme.php#L1155) to ignore the `/optional` directory

**The Development gulp Process (`npm run dev`) Does The Following:**
- [ ] Compiles unminified CSS from `/css/src/` to `/css/` passing through PostCSS and StyleLint
- [ ] Compiles unminified JS from `/js/src/` to `/js/` passing through Babel and ESLint
- [ ] Checks PHP files against PHPCS
- [ ] Watches for changes to PHP files, JS and CSS and re-runs their corresponding build task
- [ ] Initializes BrowserSync
- [ ] Hot reloads/refreshes BrowserSync whenever PHP, JS, CSS or image files change

**The Production gulp Process (`npm run bundle`) Does The Following:**
- [ ] Compiles unminified CSS from `/css/src/` to `/css/` passing through PostCSS and StyleLint
- [ ] Compiles unminified JS from `/js/src/` to `/js/` passing through Babel and ESLint
- [ ] Checks PHP files against PHPCS
- [ ] Copy files from the dev theme directory to a production theme directory
    - [ ] Only copy files based on a whitelist defined in `/config/themeConfig.js`.
    - [ ] The whitelist defaults to: all PHP files (excluding the `optional` directory), the `css` and `js`directories (excluding their `src` subdirectories) and the `images` directory
- [ ] Runs prefix renaming process on all PHP and JS files in the production theme directory
- [ ] Runs image optimization process on `/images/` folder in the production theme directory
- [ ] Runs translate processes on the production theme directory
- [ ] Creates a `.zip` archive of the production theme directory
    - [ ] Can be disabled with a setting in `/config/themeConfig.json`
- [ ] The production build directory can be disabled with a setting in `/config/themeConfig.json`. In this case, a temporary directory will be used to create the `.zip` then deleted after the production build process


Am I missing anything?

**Comment by mor10** [comment]
Missing in production process:
- `readme.txt`
- `screenshot.png`

**Comment by ataylorme** [comment]
By missing in the process do you simply mean they should be on the whitelist to be copied or they should be generated/altered?

**Comment by mor10** [comment]
copied. Otherwise they'll be missing in the production folder

**Comment by mor10** [comment]
One consideration: The `/pluggable/` folder will contain JS and CSS files (already does). These need to be handled as well.

**Comment by ataylorme** [comment]
The checklist has been updated so the whitelist includes `readme.txt`, `screenshot.png` and `/pluggable` for copying

**Comment by muggeridge** [comment]
Will the Lynda course be upddated to reflect these updates? As someone who loves learning Morten's courses on Lynda it would be re-assuring to know I could learn these new WPRig build processes.

**Comment by ataylorme** [comment]
#133 started. I have the file restructuring done. I will work on the gulp tasks as I have time. A review of the moved files would be appreciated.

**Comment by andreibarbuta** [comment]
This idea is really great! Having a dev theme to work on and a build theme to be exported to either folder or zip ( by option ) is best. This would make a world of difference for pushing changes to a live site since you would only need to push the build theme and ignore the dev theme.
Also excluding php templates ( pages, posts, etc... ) would eliminate the duplication issues in the UI.
I would also like to add that SASS would benefit immensely since it will load sourcemap files in development and production files on live server ( by exporting only the build theme ).
Question: Should SASS also have CSS Nano applied, or the resulting css be processed like any other css file when moved to build?

**Comment by ataylorme** [comment]
Closed by #133. @andreibarbuta we would love help with testing [`v2.0-alpha`](https://github.com/wprig/wprig/tree/v2.0-alpha)
